File: strings/08_valid_word_abbreviation/valid_word_abbreviation.py
class Solution:
    def valid_word_abbreviation(self, word: str, abbr: str) -> bool:
        if not self._check_constraints(word, abbr):
            raise ValueError("Constraints violated")

        cnt_id = -1
        int_ch = -1

        for i in range(len(abbr)):
            ascii = ord(abbr[i])
            if (ascii >= 48 and ascii <= 57):
                # If there is a leading 0 anywhere in 'abbr', return False
                if int_ch == -1 and (ascii - 48) == 0:
                    return False
                int_ch *= 0 if int_ch == -1 else 10
                int_ch += ascii - 48
                # If a number in 'abbr' is greater than the length of 'word', return False
                if int_ch > len(word):
                    return False
            else:
                cnt_id += 0 if int_ch == -1 else int_ch
                int_ch = -1
                cnt_id += 1
                if word[cnt_id] != abbr[i]:
                    return False
            if i == len(abbr) - 1:
                cnt_id += 0 if int_ch == -1 else int_ch
                if cnt_id != len(word) - 1:
                    return False

        return True

    # Helper function to check the constraints
    def _check_constraints(self, w: str, a: str) -> bool:
        if len(w) < 1 or len(w) > 20:
            return False
        if len(a) < 1 or len(a) > 20:
            return False
        _int = 0
        for ch in a:
            ascii = ord(ch)
            if not ((ascii >= 97 and ascii <= 122) or (ascii >= 48 and ascii <= 57)):
                return False
            else:
                if (ascii >= 48 and ascii <= 57):
                    _int *= 10
                    _int += ascii
                    if _int > 2**16 - 1:
                        return False
                else:
                    if _int > 2**16 - 1:
                        return False
                    _int = 0

        return True

File: strings/08_valid_word_abbreviation/valid_word_abbreviation.cpp
#include <iostream>
#include <string>
#include <cmath>
#include <stdexcept>
#include <cstdint>

using namespace std;


class Solution {
public:
    bool validWordAbbreviation(const string& word, const string& abbr) {
        if (!checkConstraints(word, abbr)) {
            throw std::runtime_error("Constraints violated");
        }
        int cnt_id = -1;
        int int_ch = -1;

        for (int i = 0; i < abbr.size(); ++i) {
            int ascii = static_cast<int>(abbr[i]);
            if (ascii >= 48 && ascii <= 57) {
                // If there is a leading 0 anywhere in 'abbr', return false
                if (int_ch == -1 && (ascii - 48) == 0) return false;
                int_ch *= int_ch == -1 ? 0 : 10;
                int_ch += ascii - 48;
                // If a number in 'abbr' is greater than the length of 'word', return false
                if (int_ch > word.size()) return false;
            } else {
                cnt_id += int_ch == -1 ? 0 : int_ch;
                int_ch = -1;
                cnt_id++;
                if (word[cnt_id] != abbr[i]) return false;
            }
            if (i == abbr.size() - 1) {
                cnt_id += int_ch == -1 ? 0 : int_ch;
                if (cnt_id != word.size() - 1) return false;
            }  
        }
    return true;
    }

private:
    // Helper function to check the constraints
    bool checkConstraints(const string& w, const string& a) {
        if (w.size() < 1 || w.size() > 20) return false;
        if (a.size() < 1 || a.size() > 20) return false;

        for (auto& ch : w) {
            int ascii = static_cast<int>(ch);
            if (ch < 97 || ch > 122) {
                return false;
            }
        }
        int _int = 0;
        for (auto& ch : a) {
            int ascii = static_cast<int>(ch);
            if (!((ascii >= 97 && ascii <= 122) || (ascii >= 48 && ascii <= 57))) {
                return false;
            } else {
                if (ascii >= 48 && ascii <= 57) {
                    _int *= 10;
                    _int += ascii;
                    if (_int > INT32_MAX) {
                        return false;
                    }
                } else {
                    if (_int > INT32_MAX) {
                        return false;
                    }
                    _int = 0;
                }
            }
        }
        return true;
    }
};
